src/AudioToolbox/AudioType.cs

```diff
@@ -1484,6 +1484,70 @@ public override string ToString ()
 		}
 	}
 
+	/// <summary>This struct represents the native <see href="https://developer.apple.com/documentation/coreaudiotypes/audiobufferlist">AudioBufferList</see> struct.</summary>
+	/// <remarks>
+	///   <para>
+	///     Typically it's better to use the <see cref="AudioBuffers" /> class to wrap a pointer to a native AudioBufferList,
+	///     but some audio code needs to minimize memory allocations due to being executed in a realtime thread. In that case,
+	///     using this struct is better, because it's possible to use it without incurring any memory allocations.
+	///   </para>
+	///
+	///   <para>
+	///     Note that this struct should never be created in C#, the only valid way to use it is to cast a pointer (<see cref="IntPtr" />)
+	///     to a pointer of this struct:
+	///   </para>
+	///
+	///   <example>
+	///     <code lang="csharp lang-csharp"><![CDATA[
+	/// public unsafe static void Callback (IntPtr audioBufferListPtr) {
+	///     AudioBufferList* audioBufferList = (AudioBufferList* ) audioBufferListPtr;
+	///     for (var i = 0; i < audioBufferList->Count; i++) {
+	///         AudioBuffer* buffer = audioBufferList->GetBuffer (index),
+	///         // Use the buffer for something
+	///     }
+	/// }
+	///   ]]></code>
+	///   </example>
+	/// </remarks>
+#if NET
+	[SupportedOSPlatform ("ios")]
+	[SupportedOSPlatform ("maccatalyst")]
+	[SupportedOSPlatform ("macos")]
+	[SupportedOSPlatform ("tvos")]
+#endif
+	[StructLayout (LayoutKind.Sequential)]
+	public unsafe readonly ref struct AudioBufferList {
```

Comment:
Would it be possible to make this struct non-creatable by adding a private constructor? Haven't tested if it can still be cast so just curious.

Reply:
Is it possible for us to implement a public ctor that throws and exception in DEBUG builds, but is optimized out for Release builds?

Reply:
> Is using `ref` appropriate here, it constrains the usage of this struct pretty heavily, especially around async/await and lambdas.

I can't think of any example where it would cause a problem, having in mind that this struct is supposed to be used as a pointer only.

Also it's easier to "unref" later on if we find out we were wrong than it is to add it back.

> Is it possible for us to implement a public ctor that throws and exception in DEBUG builds, but is optimized out for Release builds?

Technically yes, with some help from the linker.

That said, using this struct in the wrong way is still safe, because if you create it like this:

```cs
var abl = new AudioBufferList ();
```

you'd get an AudioBufferList with 0 buffers (`Count` would return 0, and `GetBuffer` would always throw an IndexOutOfRangeException).

So I think going through a lot of trouble to add a public ctor that throws an exception for DEBUG builds is a bit more work than this problem warrants (also because a struct's public ctor won't always be called).